guides/source/tuning_perf_for_deployment.md

```diff
@@ -0,0 +1,164 @@
+
+[CRuby](https://www.ruby-lang.org/en/) has a [Global Interpreter Lock](https://en.wikipedia.org/wiki/Global_interpreter_lock), often called the GVL or GIL. The GVL prevents multiple threads from running Ruby code at the same time in a single process. Multiple threads can be waiting on network data, database operations or some other non-Ruby work, but only one can actively run Ruby code at a time. This means thread-based concurrency is more efficient for applications that use a lot of I/O such as database operations or network APIs. The more I/O your application uses, the more threads it would benefit from.
+
+With the GVL, using a lot of threads has diminishing returns. A Rails app rarely benefits from more than 6. To have a large number of workers, some other concurrency method should be used.
+
+Threads are less resilient than processes. Certain errors like segmentation faults can destroy the entire process and all threads inside. A single request allocating a lot of memory can stop all threads while the garbage collector runs.
+
+### Hybrid Concurrency
+
+Puma allows forking multiple processes, each of which uses multiple threads. This provides a compromise between process-based and thread-based concurrency. Using multiple threads per process helps memory usage. Multiple processes permit running more Ruby code at the same time since there is one GVL per process.
+
+Hybrid concurrency limits the damage from a segmentation fault or other error that kills a process. A single process dying will kill that process's threads but not threads in other processes. Each process has its own independent garbage collector.
+
+Choosing Default Settings
+-------------------------
+
+Rails' default settings aren't appropriate for all application sizes. You can improve performance for your large application that serves a lot of requests by changing settings.
+
+This section contains common sense defaults based on the type and size of your application and the hosts on which it runs. You can improve performance more by testing your application specifically. See "Performance Testing" for details.
+
+These are production recommendations. Your development application will have different needs from a production application, and should use a different configuration.
+
+[Puma's deployment documentation](https://github.com/puma/puma/blob/master/docs/deployment.md) may also be useful.
+
+### Threads Per Process
+
+Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Discourse also executes many database queries per request and frequently uses Redis. More self-contained applications with fewer database and API queries benefit from around 3 threads per process.
```

Comment:
I disagree here. A well optimized Rails application shouldn't be IO heavy. A request should only do a handful of well indexed DB queries (single digit millisecond) and a handful of cache reads.

Any substantially slow IO should be done from a background job, as a result web request shouldn't spend more than half their time on IO.

Reply:
> A well optimized Rails application shouldn't be IO heavy

There are a lot of applications that mostly exist to integrate multiple I/O sources (Redis for security, DB for data, MemCache/Redis for caching.) Reporting is often like that, for instance. Some reporting is CPU-heavy, but some really isn't. You could argue that such applications shouldn't be written in Rails, I guess. But there are sometimes still good reasons to use Rails even in cases like that. For instance, you don't switch to Node.js just because it's a theoretically better architectural match if your team doesn't know it, and so sometimes you wind up with something very shim-server-like in Rails anyway.

You point out that it's mostly about the IO-to-CPU ratio. I agree. But some apps aren't particularly CPU-heavy by nature. And good caching tends to tip the balance even further toward I/O.